File: SIC_Backend/Repositories/PlacesRepository.cs
using Microsoft.EntityFrameworkCore;
using SIC_Backend.Data;
using SIC_Backend.Data.DTOs;
using SIC_Backend.Data.Models;

namespace SIC_Backend.Repositories
{
    public class PlacesRepository(ApplicationDbContext dbContext) : IPlacesRepository
    {
        public PlaceDTO GetPlaceById(int id)
        {
            var place = dbContext.Places.Where(p => p.Id == id)
                .Include(p => p.Sections)
                .ThenInclude(s => s.Devices)
                .Select(PlaceDTO.FromPlace)
                .First();

            return place;
        }

        public IEnumerable<PlaceDTO> GetPlacesByUserId(string userId)
        {
            return dbContext.Places.Where(p => p.UserId == userId)
                .Include(p => p.Sections)
                .ThenInclude(s => s.Devices)
                .Select(PlaceDTO.FromPlace)
                .ToList();
        }

        public async Task CreatePlaceAsync(CreatePlaceModel model)
        {
            var place = Place.FromCreateModel(model);

            dbContext.Places.Add(place);

            await dbContext.SaveChangesAsync();
        }

        public async Task UpdatePlaceAsync(int id, UpdatePlaceModel model)
        {
            var place = await dbContext.Places.FindAsync(id);

            if (place == null)
            {
                //throw new NotFoundException("Place not found");
            }

            place!.Name = model.Name;

            await dbContext.SaveChangesAsync();
        }

        public async Task DeletePlaceAsync(int id)
        {
            var place = await dbContext.Places.FindAsync(id);

            if (place == null)
            {
                //throw new NotFoundException("Place not found");
            }

            dbContext.Places.Remove(place!);
            await dbContext.SaveChangesAsync();
        }
    }
}

File: sic_home/lib/repositories/places_repository.dart
import 'package:dio/dio.dart';
import 'package:sic_home/apis/places_api.dart';
import 'package:sic_home/models/create_place_model.dart';
import 'package:sic_home/models/place.dart';
import 'package:sic_home/models/update_place_model.dart';
import 'package:sic_home/repositories/interfaces/i_places_repository.dart';

class PlacesRepository implements IPlacesRepository {
  final PlacesApi _placesApi = PlacesApi(Dio());

  @override
  Future<void> createPlace(CreatePlaceModel place) async {
    return await _placesApi.createPlace(place);
  }

  @override
  Future<void> deletePlace(int id) async {
    return await _placesApi.deletePlace(id);
  }

  @override
  Future<void> updatePlace(int id, UpdatePlaceModel place) async {
    return await _placesApi.updatePlace(id, place);
  }

  @override
  Future<Place> getPlaceById(int id) async {
    return await _placesApi.getPlace(id);
  }

  @override
  Future<List<Place>> getPlacesByUser(String userId) async {
    return await _placesApi.getPlacesByUser(userId);
  }
}
